auth/store_test.go

```diff
@@ -0,0 +1,150 @@
+// Copyright 2016 CoreOS, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package auth
+
+import (
+	"os"
+	"strings"
+	"testing"
+
+	"github.com/coreos/etcd/auth/authpb"
+	pb "github.com/coreos/etcd/etcdserver/etcdserverpb"
+	"github.com/coreos/etcd/mvcc/backend"
+)
+
+func TestUserAddAndAuthenticate(t *testing.T) {
+	b, tPath := backend.NewDefaultTmpBackend()
+	defer func() {
+		b.Close()
+		os.Remove(tPath)
+	}()
+
+	as := NewAuthStore(b)
+	ua := &pb.AuthUserAddRequest{Name: "foo", Password: "bar"}
+	_, err := as.UserAdd(ua)
+	if err != nil {
+		t.Fatal(err)
+	}
+	_, err = as.UserAdd(ua)
+	if err != nil {
+		if err != ErrUserAlreadyExist {
+			t.Fatalf("expected %v, got %v", ErrUserAlreadyExist, err)
+		}
+	} else {
+		t.Fatalf("expected %v, got %v", ErrUserAlreadyExist, err)
+	}
+
+	_, err = as.Authenticate("foo", "bar")
+	if err != nil {
+		t.Fatal(err)
+	}
+	_, err = as.Authenticate("foo", "")
+	if err != nil {
+		if !strings.Contains(err.Error(), "auth: authentication failed, invalid user ID or password") {
+			t.Fatalf("expected 'auth: authentication failed, invalid user ID or password' error, got %v", err)
+		}
+	} else {
+		t.Fatalf("expected 'invalid password...' error, got %v", err)
+	}
+}
+
+func TestUserDelete(t *testing.T) {
+	b, tPath := backend.NewDefaultTmpBackend()
+	defer func() {
+		b.Close()
+		os.Remove(tPath)
+	}()
+
+	as := NewAuthStore(b)
+
+	ua := &pb.AuthUserAddRequest{Name: "foo"}
+	_, err := as.UserAdd(ua)
+	if err != nil {
+		t.Fatal(err)
+	}
+	ud := &pb.AuthUserDeleteRequest{Name: "foo"}
+	_, err = as.UserDelete(ud)
+	if err != nil {
+		t.Fatal(err)
+	}
+	_, err = as.UserDelete(ud)
+	if err != nil {
+		if err != ErrUserNotFound {
+			t.Fatalf("expected %v, got %v", ErrUserNotFound, err)
+		}
+	} else {
+		t.Fatalf("expected %v, got %v", ErrUserNotFound, err)
+	}
+}
+
+func TestUserChangePassword(t *testing.T) {
```

Comment:
`change existing user with wrong password` cannot be tested with current implementation. It just overwrites with new password. It doesn't check if it provides the correct previous password.